File: go_api/catalog.go
package main

import (
	"context"
	"net/http"
	"time"

	"github.com/labstack/echo/v4"
	"go.mongodb.org/mongo-driver/bson"
)

type Catalog struct {
	Owner   string         `json:"owner"`
	Catalog []CatalogEntry `json:"catalog"`
}

type CatalogEntry struct {
	CategoryTitle       string               `json:"categoryTitle"`
	CategoryDescription string               `json:"categoryDescription"`
	Items               []StreamlinedProduct `json:"items"`
}
type StreamlinedProduct struct {
	Title       string `json:"title"`
	Description string `json:"description"`
	Price       int    `json:"price"`
}

func GetCatalog(c echo.Context) error {
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	owner := c.QueryParam("owner")

	category_collection := CLIENT.Database(DATABASE).Collection(category_collection)
	product_collection := CLIENT.Database(DATABASE).Collection(product_collection)

	cursor, err := category_collection.Find(ctx, bson.M{"owner": owner})
	if err != nil {
		panic(err)
	}

	catalog := Catalog{Owner: owner}

	for cursor.Next(ctx) {
		var category Category

		err = cursor.Decode(&category)
		if err != nil {
			panic(err)
		}

		catalog_entry := CatalogEntry{CategoryTitle: category.Title, CategoryDescription: category.Description}

		cursor2, err := product_collection.Find(ctx, bson.M{"category": category.Title})
		if err != nil {
			panic(err)
		}

		for cursor2.Next(ctx) {
			var product Product

			err = cursor2.Decode(&product)
			if err != nil {
				panic(err)
			}

			catalog_entry.Items = append(catalog_entry.Items, StreamlinedProduct{Title: product.Title, Description: product.Description, Price: product.Price})
		}

		catalog.Catalog = append(catalog.Catalog, catalog_entry)
	}

	return c.JSON(http.StatusOK, catalog)
}

File: api/src/catalog.rs
use std::sync::Arc;

use crate::{
    category::{Category, CATEGORY_COLLECTION},
    error::AppError,
    product::{Product, PRODUCT_COLLECTION},
    Clients, DB_NAME,
};
use axum::routing::get;
use axum::{debug_handler, Json};
use axum::{
    extract::{Query, State},
    Router,
};
use futures::TryStreamExt;
use mongodb::bson::doc;
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize)]
pub(crate) struct Catalog {
    owner: String,
    catalog: Vec<CatalogEntry>,
}

#[derive(Serialize, Deserialize)]
pub(crate) struct CatalogEntry {
    category_title: String,
    category_description: String,
    items: Vec<StreamlinedProduct>,
}

#[derive(Serialize, Deserialize)]
pub(crate) struct Owner {
    owner: String,
}

#[derive(Serialize, Deserialize)]
pub(crate) struct StreamlinedProduct {
    title: String,
    description: String,
    price: u32,
}

impl From<Product> for StreamlinedProduct {
    fn from(product: Product) -> Self {
        Self {
            title: product.title,
            description: product.description,
            price: product.price,
        }
    }
}

#[debug_handler]
pub(crate) async fn catalog(
    State(client): State<Arc<Clients>>,
    Query(owner): Query<Owner>,
) -> Result<Json<Catalog>, AppError> {
    let db = client.mongo_client.database(DB_NAME);
    let Owner { owner } = owner;

    let product_collection = db.collection::<Product>(PRODUCT_COLLECTION);
    let category_collection = db.collection::<Category>(CATEGORY_COLLECTION);

    let filter = doc! {"owner": &owner};

    let mut category_cursor = category_collection.find(filter, None).await?;
    let mut categories = Vec::new();

    while let Some(category) = category_cursor.try_next().await? {
        categories.push(category);
    }

    let mut catalog = Vec::new();

    for category in &categories {
        let filter = doc! {"owner": &owner, "category": &category.title};

        let mut product_cursor = product_collection.find(filter, None).await?;
        let mut items = Vec::new();

        while let Some(product) = product_cursor.try_next().await? {
            items.push(product.into());
        }

        catalog.push(CatalogEntry {
            category_title: category.title.clone(),
            category_description: category.description.clone(),
            items,
        });
    }

    Ok(Json(Catalog { owner, catalog }))
}

pub(crate) fn get_catalog_routes(clients: Arc<Clients>) -> Router {
    Router::new().route("/", get(catalog)).with_state(clients)
}
